### ISSUE TITLE
zodb assumes the transaction description is an ascii string

### ISSUE THREAD

**Issue by bochecha** [zopefoundation/ZODB#40]
I'm writing a Pyramid application, using ZODB 4.2.0 for the database and pyramid_tm to manage transactions.

In case it matters, I'm using Python 3.4.

pyramid_tm sets the description of a transaction to the `path_info` of the request.

For example, if a client sends a `GET /foo/bar` request, then the transaction description will be set to the string `/foo/bar`.

Then ZODB, in its `FileStorage.format:TxnHeader.asString` method, uses its `utils.as_bytes` function to encode a few strings related to the transaction, among which the description.

And `utils.as_bytes` uses the `ascii` encoder.

This all breaks down when the request `path_info` is not made up only of ASCII characters.

In many web applications, `path_info` can be any sort of things. In mine, I can even have Chinese characters, for example a `GET /森` request.

So it seems ZODB shouldn't assume everything can be encoded as ASCII.

**Comment by bochecha** [comment]
Here is a simple reproducer script:

```python
from persistent.mapping import PersistentMapping

import transaction

from ZODB.DB import DB
from ZODB.FileStorage import FileStorage


def get_app_root():
    fs = FileStorage('Data.fs')
    db = DB(fs)
    conn = db.open()
    db_root = conn.root()

    with transaction.manager:
        app_root = PersistentMapping()
        db_root['app_root'] = app_root

    return db_root['app_root']


if __name__ == '__main__':
    app_root = get_app_root()

    with transaction.manager as t:
        t.note('Something non-ascii like 森')
        app_root['item'] = PersistentMapping()
```

**Comment by mgedmin** [comment]
Are you running the reproducer script on Python 2 or 3?

**Comment by bochecha** [comment]
I'm running it on Python 3.4.2

**Comment by mgedmin** [comment]
Thank you.

**Comment by jimfulton** [comment]
This is due to ambiguity in the transaction API.

See https://github.com/zopefoundation/transaction/issues/26

When that's fixed, we'll fix ZODB accordingly.  We'll encode as utf-8.

**Comment by jimfulton** [comment]
This is fixed in ZODB 5.1.0.
